egui/src/input_state/touch_state.rs

```diff
@@ -0,0 +1,288 @@
+                distance: 0.,
+                direction: 0.,
+            };
+            let num_touches_rezip = 1. / num_touches as f32;
+
+            // first pass: calculate force, and center position:
+            for touch in self.active_touches.values() {
+                avg.force += touch.force;
+                avg.pos.x += touch.pos.x;
+                avg.pos.y += touch.pos.y;
+            }
+            avg.force *= num_touches_rezip;
+            avg.pos.x *= num_touches_rezip;
+            avg.pos.y *= num_touches_rezip;
+
+            // second pass: calculate distances from center:
+            for touch in self.active_touches.values() {
+                avg.distance += avg.pos.distance(touch.pos);
+            }
+            avg.distance *= num_touches_rezip;
+
+            // Calculate the direction from the first touch to the center position.
+            // This is not the perfect way of calculating the direction if more than two fingers
+            // are involved, but as long as all fingers rotate more or less at the same angular
+            // velocity, the shortcomings of this method will not be noticed.  One can see the
+            // issues though, when touching with three or more fingers, and moving only one of them
+            // (it takes two hands to do this in a controlled manner).  A better technique would be
+            // to store the current and previous directions (with reference to the center) for each
+            // touch individually, and then calculate the average of all individual changes in
+            // direction.  But this approach cannot be implemented locally in this method, making
+            // everything a bit more complicated.
+            let first_touch = self.active_touches.values().next().unwrap();
+            let direction_vec = (avg.pos - first_touch.pos).normalized();
+            avg.direction = direction_vec.y.atan2(direction_vec.x);
+
+            Some(avg)
+        }
+    }
+}
+struct TouchAverages {
+    pos: Pos2,
+    force: f32,
+    distance: f32,
+    direction: f32,
+}
+
+impl TouchState {
+    pub fn ui(&self, ui: &mut crate::Ui) {
+        ui.label(format!("{:?}", self));
+    }
+}
+
+impl Debug for TouchState {
+    // This outputs less clutter than `#[derive(Debug)]`:
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for (id, touch) in self.active_touches.iter() {
+            f.write_fmt(format_args!("#{:?}: {:#?}\n", id, touch))?;
+        }
+        f.write_fmt(format_args!("gesture: {:#?}\n", self.gesture_state))?;
+        Ok(())
+    }
+}
+
+/// Calculate difference between two directions, such that the absolute value of the result is
+/// minimized.
+fn normalized_angle(current_direction: f32, previous_direction: f32) -> f32 {
+    let mut angle = current_direction - previous_direction;
+    angle %= TAU;
+    if angle > PI {
+        angle -= TAU;
+    } else if angle < -PI {
+        angle += TAU;
+    }
+    dbg!(angle)
```

Comment:
remove `dbg!` please!

Reply:
Oh, clippy seems to have a blind eye although `clippy::dbg_macro` is active.
Resolved by https://github.com/emilk/egui/pull/306/commits/612069a498369a0f59e3ca4259637ec9df625ccb